implementation/src/main/java/io/smallrye/config/AbstractLocationConfigSourceFactory.java

```diff
@@ -0,0 +1,237 @@
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.eclipse.microprofile.config.spi.Converter;
+
+import io.smallrye.common.annotation.Experimental;
+
+/**
+ * This {@code AbstractLocationConfigSourceFactory} allows to initialize additional config locations with the
+ * configuration {@link AbstractLocationConfigSourceFactory#SMALLRYE_LOCATIONS}.
+ * <p>
+ *
+ * Locations set in {@link AbstractLocationConfigSourceFactory#SMALLRYE_LOCATIONS} are loaded in order and from the
+ * following resources:
+ *
+ * <ol>
+ * <li>file or directory</li>
+ * <li>classpath resource</li>
+ * <li>jar resource</li>
+ * <li>http resource</li>
+ * </ol>
+ * <p>
+ *
+ * If a profile is active, the profile resource is loaded if available and the unprofiled resource is also
+ * available. This is to keep a consistent order with the unprofiled resource. Profiles are not taken into account if
+ * the location is a directory.
+ */
+@Experimental("Load additional config locations")
+public abstract class AbstractLocationConfigSourceFactory implements ConfigSourceFactory {
+    public static final String SMALLRYE_LOCATIONS = "smallrye.config.locations";
+
+    protected abstract String[] getFileExtensions();
+
+    protected abstract ConfigSource loadConfigSource(final URL url, final int ordinal) throws IOException;
+
+    protected ConfigSource loadConfigSource(final URL url) throws IOException {
+        return this.loadConfigSource(url, ConfigSource.DEFAULT_ORDINAL);
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(final ConfigSourceContext context) {
+        final ConfigValue value = context.getValue(SMALLRYE_LOCATIONS);
+
+        final List<ConfigSource> configSources = new ArrayList<>();
+        if (value.getValue() != null) {
+            List<URL> resources = newCollectionConverter(new URLConverter(), ArrayList::new).convert(value.getValue());
+            if (resources != null) {
+                for (URL url : resources) {
+                    tryFileSystem(url, configSources, context);
+                    tryClassPath(url, configSources, context);
+                    tryHttpResource(url, configSources, context);
+                }
+            }
+        }
+
+        return configSources;
+    }
+
+    private void tryFileSystem(final URL url, final List<ConfigSource> configSources, final ConfigSourceContext context) {
+        if (url.getProtocol().startsWith("file")) {
+            // Load single file and profiles
+            final Path urlPath = Paths.get(url.getFile()).toAbsolutePath().normalize();
```

Comment:
I don't think this is necessary. The path should simply be the path of the URL, which is already absolute.  Otherwise you get into problems: calling `toAbsolutePath()` actually accesses the filesystem recursively trying to locate the root path, so if you're under a symlinked directory you'll get a different path than expected.